hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/UnderConstructionBlocks.java

```diff
@@ -0,0 +1,331 @@
+      LOG.debug("UC block {} not found on {}, total is [replicas={} / blocks={}]",
+              reportedBlock, reportingNode, count, replicasByBlockId.size());
+    }
+  }
+
+  /**
+   * If the datanode goes DEAD, the Namenode makes an assumption that
+   * any write operations have failed.
+   */
+  void removeAllUcBlocksForDatanode(DatanodeDescriptor reportingNode) {
+    if (!enabled) {
+      return;
+    }
+    if (reportingNode == null) {
+      LOG.warn("Unexpected null input {}", reportingNode);
+      return;
+    }
+    try {
+      Set<Block> toRemoveFromMap = new HashSet<>();
+      Set<BlockReplica> removedReplicas = new HashSet<>();
+      for (Map.Entry<Block, Set<BlockReplica>> entry: replicasByBlockId.entrySet()) {
+        final List<BlockReplica> storedBlocks = entry.getValue().stream()
+                .filter(replica -> reportingNode.equals(replica.getDatanode()))
+                .collect(Collectors.toList());
+        entry.getValue().removeIf(replica -> reportingNode.equals(replica.getDatanode()));
+        removedReplicas.addAll(storedBlocks);
+        count -= storedBlocks.size();
+        if (entry.getValue().isEmpty()) {
+          toRemoveFromMap.add(entry.getKey());
+        }
+      }
+      for (Block remove: toRemoveFromMap) {
+        replicasByBlockId.remove(remove);
+      }
+      final String removedBlocksString = removedReplicas.stream()
+              .map(br -> br.getBlock().toString())
+              .collect(Collectors.joining(","));
+      LOG.debug("Removed [{}] UC blocks for {}, new total is [replicas={} / blocks={}]",
+              removedBlocksString, reportingNode, count, replicasByBlockId.size());
+    } catch (Exception e) {
+      // Observed during testing that exception thrown here
+      // are caught & never logged
+      LOG.error("Remove all UnderConstruction block failed for {}",
+          reportingNode, e);
+    }
+  }
+
+  /**
+   * Add an Under Construction block replicas for a given block.
+   */
+  void addUcBlock(DatanodeDescriptor reportingNode, Block reportedBlock) {
+    if (!enabled) {
+      return;
+    }
+    if (reportingNode == null || reportedBlock == null) {
+      LOG.warn("Unexpected null input {} , {}", reportingNode, reportedBlock);
+      return;
+    }
+    try {
+      Set<BlockReplica> storedReplicasForBlock;
+      if (BlockIdManager.isStripedBlockID(reportedBlock.getBlockId())) {
+        Block blkId = new Block(BlockIdManager.convertToStripedID(reportedBlock
+                .getBlockId()));
+        storedReplicasForBlock = getBlockReplicas(blkId);
+      } else {
+        reportedBlock = new Block(reportedBlock);
+        storedReplicasForBlock = getBlockReplicas(reportedBlock);
+      }
+      addUcBlockToSet(reportingNode, reportedBlock, storedReplicasForBlock);
+    } catch (Exception e) {
+      // Observed during testing that exception thrown here
+      // are caught & never logged
+      LOG.error("Add UnderConstruction block {} on {} failed",
```

Comment:
shouldn't this be warn ?

Reply:
I can make this log WARN-level if that is preferred. The unexpected exception & resulting behaviour should not cause any critical errors in the Namenode